File: 2022/day04/clojure/solution.clj
#!/usr/bin/env bb

(require '[clojure.string :as str])

(defn parse-line [line]
  "Parse a line like '2-4,6-8' into [a1 b1 a2 b2]"
  (let [[left right] (str/split line #",")
        [a1 b1] (map parse-long (str/split left #"-"))
        [a2 b2] (map parse-long (str/split right #"-"))]
    [a1 b1 a2 b2]))

(defn parse-input [filename]
  (->> (slurp filename)
       str/split-lines
       (filter (complement str/blank?))
       (map parse-line)))

(defn fully-contains? [[a1 b1 a2 b2]]
  "Check if one range fully contains the other"
  (or (and (<= a1 a2) (>= b1 b2))
      (and (<= a2 a1) (>= b2 b1))))

(defn overlaps? [[a1 b1 a2 b2]]
  "Check if ranges overlap at all"
  (and (<= a1 b2) (<= a2 b1)))

(defn part1 [pairs]
  (count (filter fully-contains? pairs)))

(defn part2 [pairs]
  (count (filter overlaps? pairs)))

(defn -main []
  (let [script-dir (-> *file*
                       (java.io.File.)
                       (.getParentFile)
                       (.getAbsolutePath))
        input-file (str script-dir "/../input.txt")
        pairs (parse-input input-file)]
    (println "Part 1:" (part1 pairs))
    (println "Part 2:" (part2 pairs))))

(-main)

File: 2022/day04/java/Solution.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Solution {

    public static void main(String[] args) throws IOException {
        String inputFile = "../input.txt";

        int part1 = 0;
        int part2 = 0;

        try (BufferedReader reader = new BufferedReader(new FileReader(inputFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty()) {
                    continue;
                }

                String[] parts = line.split(",");
                String[] left = parts[0].split("-");
                String[] right = parts[1].split("-");

                int a1 = Integer.parseInt(left[0]);
                int b1 = Integer.parseInt(left[1]);
                int a2 = Integer.parseInt(right[0]);
                int b2 = Integer.parseInt(right[1]);

                // Part 1: Check if one range fully contains the other
                if ((a1 <= a2 && b1 >= b2) || (a2 <= a1 && b2 >= b1)) {
                    part1++;
                }

                // Part 2: Check if ranges overlap at all
                if (a1 <= b2 && a2 <= b1) {
                    part2++;
                }
            }
        }

        System.out.println("Part 1: " + part1);
        System.out.println("Part 2: " + part2);
    }
}
